### PR TITLE
Better error for HRTB error from generator interior

### PR DESCRIPTION
**Pull request by jackh726**
cc #100013

This is just a first pass at an error. It could be better, and shouldn't really be emitted in the first place. But this is better than what was being emitted before.

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -297,7 +297,7 @@ dependencies = [
  "cargo-test-macro",
  "cargo-test-support",
  "cargo-util",
- "clap 4.0.15",
```
Comment (jackh726) [inline Cargo.lock:300]: Rebase error. Will fix

**Comment by jackh726** [comment]
I think that's likely as good as we'll get here. We can no longer deduplicate as many types (and have to do two folds), so it's strictly more work.

Given that this regression only happens in a single secondary benchmark, and the above statement, I'm going to mark the perf regression as triaged.

**Comment by cjgillot** [comment]
TBH, I'm not really fond of adding a span to each `BrAnon`.  It feels a bit wrong to have spans inside `Ty` data structures.  This may become a bit of the footgun when comparing bound types to one another.

Why doesn't the reported bug cause an error before region erasure?
Should we try to prove `WF(Option<I::Future<'_, '_>>)`  in order to trigger this kind of error earlier?

**Comment by jackh726** [comment]
> TBH, I'm not really fond of adding a span to each `BrAnon`. It feels a bit wrong to have spans inside `Ty` data structures. This may become a bit of the footgun when comparing bound types to one another.

I considered giving then an optional `DefId` instead, but that doesn't quite work, because we create them so late. Also can't store them in a side table somewhere, since they aren't really "unique" enough for that.

> Why doesn't the reported bug cause an error before region erasure? Should we try to prove `WF(Option<I::Future<'_, '_>>)` in order to trigger this kind of error earlier?

This is an interesting thought I hadn't considered. Thinking a bit about this, unless we construct that obligation in just the right way, it's going to be hard to trigger (and even then...). We specifically get this error because these are late-bound vars, that get turned into placeholders, that we normalize, and register a constraint for, that gets solved during borrowck. This is specifically difficult, because it's not the type being WF that's the problem, but the type being live across the await point.

I don't think I'm doing a good job putting this into words. But essentially, it *sounds* reasonable, but I don't think it will actually work. Either because we don't know that we'll emit this error (because we don't know it will live across an await point), because we can't really construct an obligation that will emit the error but not run into the same problem (since we can't e.g. use a `BrNamed` instead, since we don't have a DefId), or both.

**Comment by jackh726** [comment]
Yes, this PR should be able to be reverted once we no longer error for this. And that is likely going to come from implied bounds here. I can mention that this is a bug.

**Comment by cjgillot** [comment]
In that case, r=me after rebase.
Thanks you for going through the boilerplate here.

**Comment by nnethercote** [comment]
Is probably noise.

@rustbot label: +perf-regression-triaged

**Pull request merged**